pkg/reconciler/tenancy/initialization/apibinder/bootstrap.go

```diff
@@ -0,0 +1,276 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apibinder
+
+import (
+	"context"
+	_ "embed"
+	"errors"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/component-base/version"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy"
+	initializationutil "github.com/kcp-dev/kcp/pkg/apis/tenancy/initialization"
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	kcpscheme "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/scheme"
+	kcpinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	initializationbootstrap "github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization"
+	"github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/apis/initialization"
+	initializationclient "github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/client/clientset/versioned"
+	initializationinformer "github.com/kcp-dev/kcp/pkg/reconciler/tenancy/initialization/client/informers/externalversions"
+)
+
+var invalidCharacters = regexp.MustCompile(`[^a-z0-9\-]`)
+
+// Bootstrap installs the apis that drive the apibinder initializer and starts the
+// controller once they're ready.
+func Bootstrap(ctx context.Context, cfg *rest.Config, installInto logicalcluster.Name) error {
+	klog.Infof("Beginning APISet Initializer bootstrapping in cluster %q", installInto)
+	gr := metav1.GroupResource{
+		Group:    initialization.GroupName,
+		Resource: "apisets",
+	}
+	crd, err := initializationbootstrap.CustomResourceDefinition(gr)
+	if err != nil {
+		return fmt.Errorf("could not determine CustomResourceDefinition: %w", err)
+	}
+
+	resourceSchema, err := apisv1alpha1.CRDToAPIResourceSchema(crd, invalidCharacters.ReplaceAllString(version.Get().String(), "-")) // TODO: find a sensible thing to use here
```

Comment:
Probably need to avoid using `version` for the prefix, as that will change with every commit, won't it? Maybe hard code it to something like v1 or rev1 or ...

Reply:
Unclear - what if the CRD changes ... ?